System Status Report: Update Networking to separate SystemPluginMapper from SystemStatusMapper

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #5071 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
Currently `SystemStatusMapper` is decoding response from `GET /wp-json/wc/v3/system_status` and returns a plugin list from there since previous usage includes only information about plugins.

For #5071 we want to get all information about system status, so this PR separates the old mapper into 2 different ones: one for plugins only and one for all information.

This PR also adds new type `Environment` for system status and updates unit tests and JSON file, hence the large number of file changes. Other changes will be added in another PR to avoid bloating this PR even more!

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Make sure that CI (all unit tests) pass
- If possible, please follow the testing steps in #5341 to make sure that system plugins are still loaded correctly.

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->
N/A

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Add new coding keys for SystemStatus

itsmeichigo committed:
Add new type Environment for SystemStatus

itsmeichigo committed:
Update system status model with environment

itsmeichigo committed:
Update test json for system status

itsmeichigo committed:
Separate system plugin and system status mappers

jaclync commented on Networking/Networking/Model/SystemStatus.swift:
```diff
@@ -31,5 +35,12 @@ private extension SystemStatus {
     enum CodingKeys: String, CodingKey {
         case activePlugins = "active_plugins"
         case inactivePlugins = "inactive_plugins"
+        case dropinMustUsePlugins = "dropins_mu_plugins"
+        case environment
+        case database
+        case theme
+        case settings
+        case pages
+        case postTypeCounts = "post_type_counts"
```
nit: are cases other than `environment` used to decode the response?

itsmeichigo replied:
Yes, I'll need to add them in a separate PR since this PR is too large already 😄

jaclync commented on Networking/Networking/Model/SystemStatus.swift:
```diff
@@ -1,16 +1,18 @@
 public struct SystemStatus: Decodable {
     let activePlugins: [SystemPlugin]
     let inactivePlugins: [SystemPlugin]
+    let environment: Environment
```
Is it worth making this optional, so that in case the environment isn't needed in the response (like when using `SystemPluginMapper` for plugins only) we don't have to fetch the environment field? Just a minor optimization on the response size and maybe time as well 🙂 

If the current remote isn't fetching `environment` field yet, it might fail to decode `SystemStatus` in the use case of `SystemPluginMapper`?

itsmeichigo replied:
Thank you for noticing this! Yes for requests for system plugins we're using filter to fetch plugins only, so this required environment can crash the request. I have updated this in b351c85e1e67470286a66431282c74f5795ee9f2.

jaclync commented on Networking/NetworkingTests/Mapper/SystemPluginMapperTests.swift:
```diff
@@ -1,9 +1,9 @@
+/// SystemPluginMapper Unit Tests
 ///
-class SystemStatusMapperTests: XCTestCase {
+final class SystemPluginMapperTests: XCTestCase {
```
nit: some unit tests for `SystemStatusMapper` would be helpful, maybe in a separate PR!

itsmeichigo replied:
Yes, that's my intention. Thank you 😄

jaclync reviewed (commented):
Mostly a question about the `environment` field! Otherwise LGTM 👍

itsmeichigo committed:
Simplify environment model

itsmeichigo committed:
Make environment optional in system status

itsmeichigo committed:
Add new json for system status response with plugins only and update system plugin mapper test with the new json

itsmeichigo commented:
Thank you @jaclync for the review! I have made a few more changes:
- Simplified `Environment` entity to use the built-in encoding behavior instead of explicit constructors - to keep code concise.
- Made `environment` property in `SystemStatus` optional to avoid crash when request to `GET /wp-json/wc/v3/system_status` has `_fields` applied.
- Add a separate JSON for test response for case where only plugins are returned from the endpoint, and updated `SystemPluginMapperTests` accordingly.

This PR is ready for another look!

jaclync reviewed (approved):
thanks for the updates, LGTM! 🚢
